File: ios/Sources/RoadSenseNSBootstrap/Pipeline/PotholeDetector.swift
import Foundation

public struct PotholeCandidate: Equatable, Sendable, Codable {
    public let latitude: Double
    public let longitude: Double
    public let magnitudeG: Double
    public let timestamp: TimeInterval

    public init(
        latitude: Double,
        longitude: Double,
        magnitudeG: Double,
        timestamp: TimeInterval
    ) {
        self.latitude = latitude
        self.longitude = longitude
        self.magnitudeG = magnitudeG
        self.timestamp = timestamp
    }
}

public struct PotholeDetector: Sendable {
    public struct Snapshot: Equatable, Sendable, Codable {
        public let spikeThresholdG: Double
        public let dipThresholdG: Double
        public let historyWindowSize: Int
        public let dipLookbackSampleCount: Int
        public let history: [Double]

        public init(
            spikeThresholdG: Double,
            dipThresholdG: Double,
            historyWindowSize: Int,
            dipLookbackSampleCount: Int,
            history: [Double]
        ) {
            self.spikeThresholdG = spikeThresholdG
            self.dipThresholdG = dipThresholdG
            self.historyWindowSize = historyWindowSize
            self.dipLookbackSampleCount = dipLookbackSampleCount
            self.history = history
        }
    }

    // Defaults tuned for typical NS driving in a passenger car: a moderate
    // pothole strike at 50–80 km/h registers ~0.8–1.6G, deep ones 2G+. The
    // dip-then-spike pattern keeps false positives down (smooth speed bumps
    // rise without dipping first).
    public var spikeThresholdG: Double = 1.0
    public var dipThresholdG: Double = -0.3
    public var historyWindowSize: Int = 50
    public var dipLookbackSampleCount: Int = 5

    private var history: [Double] = []

    public init() {}

    public init(snapshot: Snapshot) {
        self.spikeThresholdG = snapshot.spikeThresholdG
        self.dipThresholdG = snapshot.dipThresholdG
        self.historyWindowSize = snapshot.historyWindowSize
        self.dipLookbackSampleCount = snapshot.dipLookbackSampleCount
        self.history = snapshot.history
    }

    public mutating func ingest(
        verticalAccelerationG: Double,
        currentLocation: LocationSample
    ) -> PotholeCandidate? {
        history.append(verticalAccelerationG)

        if history.count > historyWindowSize {
            history.removeFirst(history.count - historyWindowSize)
        }

        guard verticalAccelerationG > spikeThresholdG else {
            return nil
        }

        let lookback = history.dropLast().suffix(dipLookbackSampleCount)
        guard let minRecent = lookback.min(), minRecent < dipThresholdG else {
            return nil
        }

        return PotholeCandidate(
            latitude: currentLocation.latitude,
            longitude: currentLocation.longitude,
            magnitudeG: verticalAccelerationG,
            timestamp: currentLocation.timestamp
        )
    }

    public func snapshot() -> Snapshot {
        Snapshot(
            spikeThresholdG: spikeThresholdG,
            dipThresholdG: dipThresholdG,
            historyWindowSize: historyWindowSize,
            dipLookbackSampleCount: dipLookbackSampleCount,
            history: history
        )
    }
}

File: android/core-sensor/src/main/kotlin/ca/roadsense/ns/sensor/PotholeDetector.kt
package ca.roadsense.ns.sensor

data class PotholeCandidate(
    val latitude: Double,
    val longitude: Double,
    val magnitudeG: Double,
    val timestamp: Double,
)

class PotholeDetector(
    var spikeThresholdG: Double = 1.0,
    var dipThresholdG: Double = -0.3,
    var historyWindowSize: Int = 50,
    var dipLookbackSampleCount: Int = 5,
) {
    data class Snapshot(
        val spikeThresholdG: Double,
        val dipThresholdG: Double,
        val historyWindowSize: Int,
        val dipLookbackSampleCount: Int,
        val history: List<Double>,
    )

    private val history: ArrayDeque<Double> = ArrayDeque()

    constructor(snapshot: Snapshot) : this(
        spikeThresholdG = snapshot.spikeThresholdG,
        dipThresholdG = snapshot.dipThresholdG,
        historyWindowSize = snapshot.historyWindowSize,
        dipLookbackSampleCount = snapshot.dipLookbackSampleCount,
    ) {
        history.addAll(snapshot.history)
    }

    fun ingest(verticalAccelerationG: Double, currentLocation: LocationSample): PotholeCandidate? {
        history.addLast(verticalAccelerationG)
        while (history.size > historyWindowSize) history.removeFirst()

        if (verticalAccelerationG <= spikeThresholdG) return null

        // dropLast(1).takeLast(dipLookbackSampleCount) — same window iOS uses.
        val lookbackEnd = history.size - 1
        val lookbackStart = (lookbackEnd - dipLookbackSampleCount).coerceAtLeast(0)
        if (lookbackEnd <= lookbackStart) return null

        var minRecent = Double.POSITIVE_INFINITY
        for (i in lookbackStart until lookbackEnd) {
            if (history[i] < minRecent) minRecent = history[i]
        }
        if (minRecent >= dipThresholdG) return null

        return PotholeCandidate(
            latitude = currentLocation.latitude,
            longitude = currentLocation.longitude,
            magnitudeG = verticalAccelerationG,
            timestamp = currentLocation.timestamp,
        )
    }

    fun snapshot(): Snapshot = Snapshot(
        spikeThresholdG = spikeThresholdG,
        dipThresholdG = dipThresholdG,
        historyWindowSize = historyWindowSize,
        dipLookbackSampleCount = dipLookbackSampleCount,
        history = history.toList(),
    )
}
